### PR TITLE
164101183 invite new member

### PR DESCRIPTION
**Pull request by jkorpelin**
# Added
* Ability to invite new users by email as part of the removal of ckan.
   
# Datamodel
* New auditlog & user-token table

### CODE REVIEW

File: ote/src/clj/ote/services/register.clj [hunk 1]
```diff
@@ -0,0 +1,101 @@
+(ns ote.services.register
```
Comment (rckopteri) [inline ote/src/clj/ote/services/register.clj]: 👍

File: ote/src/clj/ote/services/register.clj [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    (user/email-valid? email)
+    (user/username-valid? username)
+    (string? name) (not (str/blank? name))))
+
+(defn- register-user! [db auth-tkt-config {:keys [username name email password token] :as form-data}]
+  (if-not (valid-registration? form-data)
+    ;; Check errors that should have been checked on the form
+    {:success? false}
+    (with-transaction db
+      (println form-data)
```
Comment (rckopteri) [inline ote/src/clj/ote/services/register.clj]: remove println

File: ote/src/clj/ote/services/transport.clj [hunk 3]
```diff
@@ -539,7 +543,19 @@
...
+                  #{::user/user-email ::user/token}
+                  {::user/operator-id operator-id
+                   ::user/expiration (op/>= (tc/to-sql-date (t/now)))})
+        invites (map
+                  (fn [invite]
+                    (-> invite
+                      (assoc :pending? true)
+                      (rename-keys {::user/user-email :email ::user/token :token})))
+                  invites)]
+    (clojure.pprint/pprint (concat users invites))
```
Comment (rckopteri) [inline ote/src/clj/ote/services/transport.clj]: Remove pretty print

File: ote/src/clj/ote/services/transport.clj [hunk 4]
```diff
@@ -556,53 +572,91 @@
...
                   ::auditlog/created-by (get-in requester [:user :id])}]
+    ;; Send email to user to inform about the new membership.
+    (try
+      (email/send!
+        email
+        {:to (:user_email updated-member)
+         :subject title
+         :body [{:type "text/html;charset=utf-8"
+                 :content (str email-template/html-header
+                            (html (email-template/notify-user-new-member updated-member requester operator title)))}]})
```
Comment (rckopteri) [inline ote/src/clj/ote/services/transport.clj]: Where title is assigned?

File: ote/src/clj/ote/services/transport.clj [hunk 5]
```diff
@@ -556,53 +572,91 @@
...
+                   {::auditlog/name "new-member-email", ::auditlog/value (str user-email)}]
+                  ::auditlog/event-timestamp (java.sql.Timestamp. (System/currentTimeMillis))
+                  ::auditlog/created-by (get-in requester [:user :id])}]
+    (try
+      (email/send!
+        email-config
+        {:to user-email
+         :subject title
+         :body [{:type "text/html;charset=utf-8"
+                 :content (html (email-template/new-user-invite requester operator title (::user/token inserted-token)))}]})
```
Comment (rckopteri) [inline ote/src/clj/ote/services/transport.clj]: You need to add email-template/html-header
to :content. And if you like you can implement that header with better way than what I came up with.

File: ote/src/cljs/ote/app/controller/login.cljs [hunk 6]
```diff
@@ -249,3 +252,28 @@
               {:on-success (tuck/send-async! ->RequestPasswordResetResponse (:email app))
                :on-failure (tuck/send-async! ->ServerError)})
   app)
+
+(define-event ValiditySuccess [response]
+  {:path [:register]}
+  (println app)
```
Comment (rckopteri) [inline ote/src/cljs/ote/app/controller/login.cljs]: Remove println

File: ote/src/cljs/ote/views/register.cljs [hunk 7]
```diff
@@ -1,15 +1,25 @@
...
             [ote.ui.buttons :as buttons]
             [ote.ui.common :as common]
+            [ote.db.user :as user]
+            [ote.ui.notification :as notification]))
 
-            [ote.db.user :as user]))
 
+(defn token-check
+  [token-info]
+  (println token-info)
```
Comment (rckopteri) [inline ote/src/cljs/ote/views/register.cljs]: Remove println

File: ote/src/cljs/ote/views/register.cljs [hunk 8]
```diff
@@ -1,15 +1,25 @@
...
+  (if (not (= token-info :token-invalid))
+    (notification/notification
+      {:text (str "Olet liittymässä osaksi palvelua " (:name token-info))
```
Comment (rckopteri) [inline ote/src/cljs/ote/views/register.cljs]: Would not hurt if you added this to translation files

File: ote/src/cljs/ote/views/register.cljs [hunk 9]
```diff
@@ -18,13 +28,16 @@
...
 ;; to be worth it for this form alone.
 
-(defn register [e! {:keys [form-data email-taken username-taken] :as register} user]
-  (let [edited (r/atom #{}) ; keep track of blurred fields
+(defn register [e! {:keys [token] :as params} {:keys [form-data email-taken username-taken token-info] :as register} user]
+  (let [edited (r/atom #{})                                 ; keep track of blurred fields
         edit! #(swap! edited conj %)]
-    (fn [e! {:keys [form-data email-taken username-taken] :as register} user]
+    (fn [e! {:keys [token] :as params} {:keys [form-data email-taken username-taken token-info] :as register} user]
+      (println "token-info register: " token-info)
```
Comment (rckopteri) [inline ote/src/cljs/ote/views/register.cljs]: Remove println

File: ote/src/cljs/ote/views/register.cljs (hunk 8 above)
Comment (jkorpelin) [reply ote/src/cljs/ote/views/register.cljs]: Jussi is going to figure out proper texts for these notifications first.

File: ote/src/clj/ote/services/transport.clj (hunk 4 above)
Comment (jkorpelin) [reply ote/src/clj/ote/services/transport.clj]: In the `let` above.

**Pull request merged**